File: src/collections_and_containers/cpp/binary_search_tree/src/binary_search_tree.cpp
#include "binary_search_tree.h"
#include <algorithm>
#include <stdexcept>

template <class T>
BinarySearchTree<T>::BinarySearchTree() : root(nullptr), n(0) {}

template <class T>
BinarySearchTree<T>::BinarySearchTree(const BinarySearchTree<T> &t)
    : root(nullptr), n(0) {
  for (const auto &v : t.in_order_traversal()) {
    insert(v);
  }
}

template <class T>
BinarySearchTree<T> &
BinarySearchTree<T>::operator=(const BinarySearchTree<T> &t) {
  if (this != &t) {
    clear();
    for (const auto &v : t.in_order_traversal()) {
      insert(v);
    }
  }
  return *this;
}

template <class T> BinarySearchTree<T>::~BinarySearchTree() { clear(); }

template <class T> void BinarySearchTree<T>::insert(const T &v) {
  insert(v, root);
}

template <class T> void BinarySearchTree<T>::remove(const T &v) {
  remove(v, root);
}

template <class T> bool BinarySearchTree<T>::contains(const T &v) const {
  return contains(v, root);
}

template <class T> T BinarySearchTree<T>::find_min() const {
  if (root == nullptr) {
    throw std::out_of_range("Tree is empty");
  }
  const Node *p = root.get();
  while (p->left != nullptr) {
    p = p->left.get();
  }
  return p->data;
}

template <class T> T BinarySearchTree<T>::find_max() const {
  if (root == nullptr) {
    throw std::out_of_range("Tree is empty");
  }
  const Node *p = root.get();
  while (p->right != nullptr) {
    p = p->right.get();
  }
  return p->data;
}

template <class T> void BinarySearchTree<T>::clear() {
  clear(root);
  n = 0;
}

template <class T> int BinarySearchTree<T>::height() const {
  return height(root);
}

template <class T> int BinarySearchTree<T>::size() const { return n; }

template <class T> bool BinarySearchTree<T>::empty() const { return n == 0; }

template <class T>
std::vector<T> BinarySearchTree<T>::in_order_traversal() const {
  std::vector<T> v;
  in_order_traversal(root, v);
  return v;
}

template <class T>
std::vector<T> BinarySearchTree<T>::pre_order_traversal() const {
  std::vector<T> v;
  pre_order_traversal(root, v);
  return v;
}

template <class T>
std::vector<T> BinarySearchTree<T>::post_order_traversal() const {
  std::vector<T> v;
  post_order_traversal(root, v);
  return v;
}

template <class T>
void BinarySearchTree<T>::insert(const T &v, std::unique_ptr<Node> &p) {
  if (p == nullptr) {
    p = std::make_unique<Node>();
    p->data = v;
    n++;
  } else if (v < p->data) {
    insert(v, p->left);
  } else if (v > p->data) {
    insert(v, p->right);
  } else {
    throw std::invalid_argument("Value already exists in tree");
  }
}

template <class T>
void BinarySearchTree<T>::remove(const T &v, std::unique_ptr<Node> &p) {
  if (p == nullptr) {
    throw std::invalid_argument("Value does not exist in tree");
  } else if (v < p->data) {
    remove(v, p->left);
  } else if (v > p->data) {
    remove(v, p->right);
  } else {
    if (p->left == nullptr && p->right == nullptr) {
      p = nullptr;
      n--;
    } else if (p->left == nullptr) {
      p = std::move(p->right);
      n--;
    } else if (p->right == nullptr) {
      p = std::move(p->left);
      n--;
    } else {
      // Find in-order successor
      Node *successor = p->right.get();
      while (successor->left != nullptr) {
        successor = successor->left.get();
      }
      p->data = successor->data;
      remove(successor->data, p->right);
    }
  }
}

template <class T>
bool BinarySearchTree<T>::contains(const T &v,
                                   const std::unique_ptr<Node> &p) const {
  if (p == nullptr) {
    return false;
  } else if (v < p->data) {
    return contains(v, p->left);
  } else if (v > p->data) {
    return contains(v, p->right);
  } else {
    return true;
  }
}

template <class T> void BinarySearchTree<T>::clear(std::unique_ptr<Node> &p) {
  if (p != nullptr) {
    clear(p->left);
    clear(p->right);
    p = nullptr;
  }
}

template <class T>
int BinarySearchTree<T>::height(const std::unique_ptr<Node> &p) const {
  if (p == nullptr) {
    return 0;
  } else {
    return 1 + std::max(height(p->left), height(p->right));
  }
}

template <class T>
void BinarySearchTree<T>::in_order_traversal(const std::unique_ptr<Node> &p,
                                             std::vector<T> &v) const {
  if (p != nullptr) {
    in_order_traversal(p->left, v);
    v.push_back(p->data);
    in_order_traversal(p->right, v);
  }
}

template <class T>
void BinarySearchTree<T>::pre_order_traversal(const std::unique_ptr<Node> &p,
                                              std::vector<T> &v) const {
  if (p != nullptr) {
    v.push_back(p->data);
    pre_order_traversal(p->left, v);
    pre_order_traversal(p->right, v);
  }
}

template <class T>
void BinarySearchTree<T>::post_order_traversal(const std::unique_ptr<Node> &p,
                                               std::vector<T> &v) const {
  if (p != nullptr) {
    post_order_traversal(p->left, v);
    post_order_traversal(p->right, v);
    v.push_back(p->data);
  }
}

template class BinarySearchTree<char>;
template class BinarySearchTree<int>;
template class BinarySearchTree<std::string>;
template class BinarySearchTree<double>;
template class BinarySearchTree<float>;

File: src/collections_and_containers/python/binary_search_tree/src/binary_search_tree.py
"""Binary Search Tree implementation."""

from __future__ import annotations

from typing import Any, Optional


class Node:
    """A node in a binary search tree."""

    def __init__(self, data: Any) -> None:
        self.data = data
        self.left: Optional[Node] = None
        self.right: Optional[Node] = None


class BinarySearchTree:
    """A binary search tree data structure."""

    def __init__(self) -> None:
        self.root: Optional[Node] = None
        self.n = 0

    def insert(self, v: Any) -> None:
        """Insert a value into the tree."""
        if self.root is None:
            self.root = Node(v)
        else:
            self._insert(v, self.root)
        self.n += 1

    def _insert(self, v: Any, p: Node) -> None:
        if v < p.data:
            if p.left is None:
                p.left = Node(v)
            else:
                self._insert(v, p.left)
        else:
            if p.right is None:
                p.right = Node(v)
            else:
                self._insert(v, p.right)

    def remove(self, v: Any) -> None:
        """Remove a value from the tree."""
        if self.root.data == v and self.root.left is None and self.root.right is None:
            self.root = None
        else:
            self._remove(v, self.root)
        self.n -= 1

    def _remove(self, v: Any, p: Optional[Node]) -> Optional[Node]:
        def _find_min(p: Optional[Node]) -> Optional[Any]:
            if p is None:
                return None
            while p.left is not None:
                p = p.left
            return p.data

        if p is None:
            return None

        if v < p.data:
            p.left = self._remove(v, p.left)
        elif v > p.data:
            p.right = self._remove(v, p.right)
        else:
            if p.left is None:
                return p.right
            elif p.right is None:
                return p.left
            else:
                p.data = _find_min(p.right)
                p.right = self._remove(p.data, p.right)
        return p

    def contains(self, v: Any) -> bool:
        """Check if the tree contains a value."""
        return self._contains(v, self.root)

    def _contains(self, v: Any, p: Optional[Node]) -> bool:
        if p is None:
            return False
        if v < p.data:
            return self._contains(v, p.left)
        elif v > p.data:
            return self._contains(v, p.right)
        else:
            return True

    def clear(self) -> None:
        """Remove all elements from the tree."""
        self.root = None
        self.n = 0

    def height(self) -> int:
        """Return the height of the tree."""
        if self.root is None:
            return 0
        return self._height(self.root)

    def _height(self, p: Optional[Node]) -> int:
        if p is None:
            return 0
        return 1 + max(self._height(p.left), self._height(p.right))

    def size(self) -> int:
        """Return the number of elements in the tree."""
        return self.n

    def empty(self) -> bool:
        """Check if the tree is empty."""
        return self.n == 0

    def in_order_traversal(self) -> list[Any]:
        """Return elements in in-order traversal (left, root, right)."""
        v: list[Any] = []
        self._in_order_traversal(self.root, v)
        return v

    def _in_order_traversal(self, p: Optional[Node], v: list[Any]) -> None:
        if p is None:
            return
        self._in_order_traversal(p.left, v)
        v.append(p.data)
        self._in_order_traversal(p.right, v)

    def pre_order_traversal(self) -> list[Any]:
        """Return elements in pre-order traversal (root, left, right)."""
        v: list[Any] = []
        self._pre_order_traversal(self.root, v)
        return v

    def _pre_order_traversal(self, p: Optional[Node], v: list[Any]) -> None:
        if p is None:
            return
        v.append(p.data)
        self._pre_order_traversal(p.left, v)
        self._pre_order_traversal(p.right, v)

    def post_order_traversal(self) -> list[Any]:
        """Return elements in post-order traversal (left, right, root)."""
        v: list[Any] = []
        self._post_order_traversal(self.root, v)
        return v

    def _post_order_traversal(self, p: Optional[Node], v: list[Any]) -> None:
        if p is None:
            return
        self._post_order_traversal(p.left, v)
        self._post_order_traversal(p.right, v)
        v.append(p.data)

    def __str__(self) -> str:
        return str(self.in_order_traversal())
